common/nymkkt/src/frame.rs

```diff
@@ -0,0 +1,146 @@
+    }
+
+    pub fn frame_length(&self) -> usize {
+        self.context.len() + self.session_id.len() + self.body.len() + self.signature.len()
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::with_capacity(self.frame_length());
+        bytes.extend_from_slice(&self.context);
+        bytes.extend_from_slice(&self.body);
+        bytes.extend_from_slice(&self.session_id);
+        bytes.extend_from_slice(&self.signature);
+        bytes
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<(Self, KKTContext), KKTError> {
+        if bytes.len() < KKT_CONTEXT_LEN {
+            Err(KKTError::FrameDecodingError {
+                info: format!(
+                    "Frame is shorter than expected context length: actual {} != expected {}",
+                    bytes.len(),
+                    KKT_CONTEXT_LEN
+                ),
+            })
+        } else {
+            let context_bytes = Vec::from(&bytes[0..KKT_CONTEXT_LEN]);
+
+            let context = KKTContext::try_decode(&context_bytes)?;
+
+            let (mut session_id, mut body, mut signature): (Vec<u8>, Vec<u8>, Vec<u8>) =
+                (vec![], vec![], vec![]);
+
+            if bytes.len() == context.full_message_len() {
+                if context.body_len() > 0 {
+                    body.extend_from_slice(
+                        &bytes[KKT_CONTEXT_LEN..KKT_CONTEXT_LEN + context.body_len()],
+                    );
+                }
+                if context.session_id_len() > 0 {
+                    session_id.extend_from_slice(
+                        &bytes[KKT_CONTEXT_LEN + context.body_len()
+                            ..KKT_CONTEXT_LEN + context.body_len() + context.session_id_len()],
+                    );
+                }
+                if context.signature_len() > 0 {
+                    signature.extend_from_slice(
+                        &bytes[KKT_CONTEXT_LEN + context.body_len() + context.session_id_len()
+                            ..KKT_CONTEXT_LEN
+                                + context.body_len()
+                                + context.session_id_len()
+                                + context.signature_len()],
+                    );
+                }
+
+                Ok((
+                    KKTFrame::new(&context_bytes, &body, &session_id, &signature),
+                    context,
+                ))
+            } else {
+                Err(KKTError::FrameDecodingError {
+                    info: format!(
+                        "Frame is shorter than expected: actual {} != expected {}",
+                        bytes.len(),
+                        context.full_message_len()
+                    ),
+                })
+            }
+        }
+    }
+}
+
+pub struct EncryptedKKTRequest {
+    ephermeral_key: PublicKey,
+    body: Vec<u8>,
+}
+impl EncryptedKKTRequest {
+    pub fn to_bytes(mut self) -> Vec<u8> {
+        self.body.extend(self.ephermeral_key.to_bytes());
```

Comment:
just the nittest of nits: I'd move `self.body` to a different binding to make it more readable. in a pre-coffee state I was confused reading this thinking: "wait, why is he mutating the body content"
```suggestion
    pub fn to_bytes(self) -> Vec<u8> {
        let mut bytes = self.body;
        bytes.extend(self.ephermeral_key.to_bytes());
        bytes
    }
```